visual_lib/src/feature_detection/harris/harris_gpu_cuda_tools.cu

```diff
@@ -0,0 +1,309 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "vilib/cuda_common.h"
+#include "vilib/feature_detection/harris/harris_gpu_cuda_tools.h"
+
+namespace vilib {
+
+static void harris_gpu_find_best_params(const int & cols,
+                                        int & threadX,
+                                        int & threadY) {
+  int t_best = 128;
+  for(int t=t_best,idle_threads_best = t_best;t>16;t>>=1) {
+    int idle_threads_now = ((cols + t-1)/t)*t - cols;
+    if(idle_threads_best > idle_threads_now) {
+      t_best = t;
+      idle_threads_best = idle_threads_now;
+    }
+  }
+  threadX = t_best;
+  threadY = 128/t_best;
+}
+
+static void harris_gpu_find_best_vector_params(const int & cols,
+                                               const int & input_pitch,
+                                               const int & output_pitch,
+                                               int & v,
+                                               int & cols_vectorized,
+                                               int & input_pitch_vectorized,
+                                               int & output_pitch_vectorized,
+                                               int & threadX,
+                                               int & threadY) {
+  // Select the most efficient vectorized version
+  v=4;
+  for(;v>0;v>>=1) {
+    int bitfield = v-1;
+    if(((cols & bitfield) != 0) ||
+       ((input_pitch & bitfield) != 0) ||
+       ((output_pitch & bitfield) != 0)) {
+         continue;
+    }
+    break;
+  }
+  cols_vectorized = cols/v;
+  input_pitch_vectorized = input_pitch/v;
+  output_pitch_vectorized = output_pitch/v;
+  // Select a good thread block size that minimizes the number of idle threads
+  int t_best = 128;
+  for(int t=t_best,idle_threads_best = t_best;t>16;t>>=1) {
+    int idle_threads_now = ((cols_vectorized + t-1)/t)*t - cols_vectorized;
+    if(idle_threads_best > idle_threads_now) {
+      t_best = t;
+      idle_threads_best = idle_threads_now;
+    }
+  }
+  threadX = t_best;
+  threadY = 128/t_best;
+}
+
+template <int N>
+__global__ void array_multiply_kernel(const float * __restrict__ a,
+                                      const float * __restrict__ b,
+                                      float * result,
+                                      const int cols,
+                                      const int rows,
+                                      const int common_pitch) {
+  int tx = blockIdx.x * blockDim.x + threadIdx.x;
```

Comment:
why not `const` for indices? also for `idx` below and in all other kernel functions...